src/stream/src/executor/hash_agg.rs

```diff
@@ -410,68 +428,48 @@ impl<K: HashKey, S: StateStore> HashAggExecutor<K, S> {
+            let futs_of_all_groups = vars
+                .group_change_set
+                .drain()
+                .map(|key| {
                     // Get agg group of the key.
-                    let agg_group = {
-                        let mut ptr: NonNull<_> = vars
-                            .agg_group_cache
-                            .get_mut(&key)
-                            .expect("changed group must have corresponding AggGroup")
-                            .into();
-                        // SAFETY: `key`s in `keys_in_batch` are unique by nature, because they're
-                        // from `group_change_set` which is a set.
-                        unsafe { ptr.as_mut() }
-                    };
-                    async {
-                        let curr_outputs = agg_group.get_outputs(&this.storages).await?;
-                        Ok::<_, StreamExecutorError>((key, agg_group, curr_outputs))
-                    }
+                    let mut ptr: NonNull<_> = vars
+                        .agg_group_cache
+                        .get_mut(&key)
+                        .expect("changed group must have corresponding AggGroup")
+                        .into();
+                    // SAFETY: `key`s in `keys_in_batch` are unique by nature, because they're
+                    // from `group_change_set` which is a set.
+                    unsafe { ptr.as_mut() }
+                })
+                .map(|agg_group| async {
+                    // Get agg outputs and build change.
+                    let curr_outputs = agg_group.get_outputs(&this.storages).await?;
+                    let change = agg_group.build_change(curr_outputs);
+                    Ok::<_, StreamExecutorError>(change)
                 });
-                let outputs_in_batch: Vec<_> = stream::iter(futs)
-                    .buffer_unordered(10)
-                    .fuse()
-                    .try_collect()
-                    .await?;
-
-                for (_key, agg_group, curr_outputs) in outputs_in_batch {
-                    if let Some(change) = agg_group.build_change(curr_outputs) {
-                        agg_group.apply_change_to_builders(&change, &mut builders, &mut new_ops);
-                        agg_group.apply_change_to_result_table(&change, &mut this.result_table);
+
+            // TODO(rc): figure out a more reasonable concurrency limit.
+            const MAX_CONCURRENT_TASKS: usize = 100;
+            let mut futs_chunks = IterChunks::chunks(futs_of_all_groups, MAX_CONCURRENT_TASKS);
+            while let Some(futs) = futs_chunks.next() {
+                let changes = futures::future::try_join_all(futs).await?;
+                for change in changes.into_iter().flatten() {
+                    apply_change_to_result_table(&change, &mut this.result_table);
+
+                    chunk_of_changes.push(change);
+                    if chunk_of_changes.len() == this.chunk_size {
```

Comment:
Only yield a chunk when number of changes reach `this.chunk_size`. This is a preparation for `emit on window close`.